app/src/main/java/ru/gorshenev/themesstyles/presentation/ui/channels/StreamFragment.kt

```diff
@@ -26,29 +28,55 @@ import ru.gorshenev.themesstyles.presentation.ui.channels.items.TopicUi
 import ru.gorshenev.themesstyles.presentation.ui.chat.ChatFragment
 import ru.gorshenev.themesstyles.utils.Utils.setDivider
 
-class StreamFragment : MvpFragment<StreamView, StreamPresenter>(R.layout.fragment_channels_stream),
-    StreamView {
+class StreamFragment : Fragment(R.layout.fragment_channels_stream),
+    MviView<StreamState, UiEffects> {
     private val binding: FragmentChannelsStreamBinding by viewBinding()
 
-    private val streamPresenter by lazy { StreamPresenter(GlobalDI.INSTANSE.streamRepository) }
+    private val streamType by lazy { arguments?.get(STR_TYPE) as StreamType }
 
-    override fun getPresenter(): StreamPresenter = streamPresenter
+    private val streamViewModel: MviViewModel<StreamAction, StreamState, UiEffects> by viewModels {
+        val streamStore: Store<StreamAction, StreamState, UiEffects> =
+            Store(
+                reducer = StreamReducer(),
+                middlewares = listOf(
+                    StreamUploadMiddleware(GlobalDI.INSTANSE.streamRepository),
+                    StreamSearchMiddleware(),
+                    StreamOnStreamClickMiddleware(),
+                    StreamOnTopicClickMiddleware()
+                ),
+                initialState = StreamState.Loading
+            )
+        MviViewModelFactory(streamStore)
+    }
 
-    override fun getMvpView(): StreamView = this
-
-    private val streamType by lazy { arguments?.get(STR_TYPE) as StreamType }
+    private val currentState
+        get() = streamViewModel.state
 
     private val holderFactory: HolderFactory = StreamsHolderFactory(
-        onStreamClick = { streamId -> getPresenter().onStreamClick(streamId) },
-        onTopicClick = { topicId -> getPresenter().onTopicClick(topicId) }
+        onStreamClick = { streamId ->
+            if (streamViewModel.state is StreamState.Result) {
+                streamViewModel.accept(
+                    StreamAction.OnStreamClick(streamId, (currentState as StreamState.Result).visibleItems)
+                )
+            }
+        },
+        onTopicClick = { topicId ->
+            if (currentState is StreamState.Result) {
```

Comment:
private fun StreamState.ifResult(action: (state: StreamState.Result) -> Unit) {
        (this as? StreamState.Result)?.let(action)
    }

Reply:
+